tlatools/org.lamport.tlatools/src/tlc2/value/impl/CharValue.java

```diff
@@ -0,0 +1,184 @@
+			else { throw e; }
+		}
+	}
+
+	@Override
+	public final int size() {
+	    try {
+	        Assert.fail("Attempted to compute the number of elements in the character " +
+	        Values.ppr(this.toString()) + ".", getSource());
+	        return 0;   // make compiler happy
+	      }
+	      catch (RuntimeException | OutOfMemoryError e) {
+	        if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
+	        else { throw e; }
+	      }
+	}
+
+	@Override
+	public final boolean isNormalized() { return true; }
+
+	@Override
+	public final Value normalize() { return this; }
+
+	@Override
+	public final boolean isDefined() { return true;	}
+
+	@Override
+	public final IValue deepCopy() { return this; }
+
+	@Override
+	public final StringBuffer toString(StringBuffer sb, int offset, boolean swallow) {
+	    try {
+	        return sb.append('\'').append(this.val).append('\'');
+	        //return sb.append('"').append(this.val).append("\"[1]");
+	      }
+	    catch (RuntimeException | OutOfMemoryError e) {
+	        if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
+	        else { throw e; }
+	      }
+	}
+
+	@Override
+	public final byte getKind() { return CHARVALUE; }
+
+	@Override
+	public final Value takeExcept(ValueExcept ex) {
+	    try {
+	        if (ex.idx < ex.path.length) {
+	          Assert.fail("Attempted to appy EXCEPT construct to the character " +
+	          Values.ppr(this.toString()) + ".", getSource());
+	        }
+	        return ex.value;
+	      }
+	      catch (RuntimeException | OutOfMemoryError e) {
+	        if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
+	        else { throw e; }
+	      }
+	}
+
+	@Override
+	public final Value takeExcept(ValueExcept[] exs) {
+	    try {
+	        if (exs.length != 0) {
+	          Assert.fail("Attempted to apply EXCEPT construct to the character " +
+	          Values.ppr(this.toString()) + ".", getSource());
+	        }
+	        return this;
+	      }
+	      catch (RuntimeException | OutOfMemoryError e) {
+	        if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
+	        else { throw e; }
+	      }
+	}
+
+	@Override
+	final boolean assignable(Value val) {    
+		try {
+			return ((val instanceof CharValue) &&
+	    	        this.val == ((CharValue)val).val);
+		} catch (RuntimeException | OutOfMemoryError e) {
+			if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
+			else { throw e; }
+		}
+	}
+
+	@Override
+	public final IValue permute(IMVPerm perm) { return this; }
+
+	@Override
+	public final long fingerPrint(long fp) {
+	    try {
+	      fp = FP64.Extend(fp, CHARVALUE) ;
+	      fp = FP64.Extend(fp, this.val) ;
+	      return fp;
+	    }
+	    catch (RuntimeException | OutOfMemoryError e) {
+	      if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
+	      else { throw e; }
+	    }
+	  }
+	
+	@Override
+	public final boolean mutates() { return false; }
```

Comment:
this is according to the doc (it normalization and fingerprinting don't mutate the object) but not in line with the e.g. IntValues...

Reply:
Adding a new basic data type to TLC introduces  problems, a serious one being  to modify the fingerprinting code to maintain the extremely small probability of collisions.   This is not trivial.  The same fingerprinting algorithm was used inside a compiler once.  A compiler bug was traced to a collision of fingerprints that should have had an astronomically small probability.   It was discovered that this was the result of a bug in the algorithm's implementation.

One possibility might be to make characters typed model values, using as type a character that is not currently allowed.

Reply:
I have no experience in this matter, but I fail to see why characters can not be handled in a similar way as integers since both are essentially numbers. In fact, it seems to me that your suggestion would have a very similar result, except for substituting the char kind byte with the model value kind byte and the character value with the unique id of the character.

Reply:
I do not understand the fingerprinting algorithm.  But I know that some smart people (researchers at one of the world's leading research labs), working from a description of the algorithm, produced an incorrect program.  This tells me that modifying code that implements the algorithm, without fully understanding the code or the algorithm, is risky.  

Fingerprinting is tricky.   You're mapping numbers, sets of number, functions on sets of numbers, etc. all into strings of bits.   For any single one of those kinds of objects, it's not too hard to find some mapping that has a small probability of collisions.  But how do you know that there is a small probability of a collision between two different kinds of value?  It's not easy.  Saying characters are like numbers so you just treat them like numbers sounds like a recipe for producing collisions between characters and numbers.

Reply:
As far as I can see, the code uses a one-way compression function that produces a 64bit tag from a tag and a byte (`FP64.Extend`). To address the problem you mentioned, every class first extends the fingerprint with a byte unique to the class, followed by the data. This leads to different sequences of bytes fed to the compression function even if values of different types share the same binary representation (at least for primitives/types of constant length).

Speaking of collisions: As far as I can tell the code of `FP64.Extend` treats chars as a single byte. Strings that contain characters outside the range 0..2^8-1 are probably rare but would result in colliding fingerprints.

Reply:
That sounds good.  I believe the problem solved by the general fingerprinting algorithm involved fingerprinting values from an unbounded set of types.  With values from a fixed set of types, the problem does seem simple.

Thanks for pointing out the problem that can arise if TLA+ is extended to allow non-ASCII characters in strings.  It would be nice if it could be documented somewhere in case that extension is ever implemented, but I don't know how to add documentation for code that no one has begun writing.